packages/nitro-protocol/contracts/MultiAssetHolder.sol

```diff
@@ -242,6 +242,255 @@ contract MultiAssetHolder is IMultiAssetHolder, StatusManager {
+                claimArgs.indices,
+                claimArgs.sourceChannelId,
+                claimArgs.sourceStateHash,
+                sourceOutcomeBytes,
+                claimArgs.targetStateHash,
+                targetOutcomeBytes
+            ); // view
+        }
+        Outcome.Allocation[] memory newSourceAllocations;
+        Outcome.Allocation[] memory newTargetAllocations;
+        Outcome.Allocation[] memory exitAllocations;
+        uint256 totalPayouts;
+        {
+            Outcome.Allocation[] memory sourceAllocations = sourceOutcome[assetIndex].allocations;
+            Outcome.Allocation[] memory targetAllocations = targetOutcome[assetIndex].allocations;
+            (
+                newSourceAllocations,
+                newTargetAllocations,
+                exitAllocations,
+                totalPayouts
+            ) = compute_claim_effects_and_interactions(
+                initialAssetHoldings,
+                sourceAllocations,
+                targetAllocations,
+                claimArgs.indexOfTargetInSource,
+                claimArgs.indices
+            ); // pure, also performs checks
+        }
+
+        _apply_claim_effects(
+            assetIndex,
+            claimArgs.sourceChannelId,
+            claimArgs.sourceStateHash,
+            sourceOutcome,
+            newSourceAllocations,
+            sourceOutcome[assetIndex].allocations[claimArgs.indexOfTargetInSource].destination, // targetChannelId
+            claimArgs.targetStateHash,
+            targetOutcome,
+            newTargetAllocations,
+            initialAssetHoldings,
+            totalPayouts
+        );
+
+        _apply_claim_interactions(targetOutcome[assetIndex], exitAllocations);
+    }
+
+    function _apply_claim_checks(
+        uint256 assetIndex,
+        uint256 indexOfTargetInSource,
+        uint256[] memory indices,
+        bytes32 sourceChannelId,
+        bytes32 sourceStateHash,
+        bytes memory sourceOutcomeBytes,
+        bytes32 targetStateHash,
+        bytes memory targetOutcomeBytes
+    )
+        internal
+        view
+        returns (
+            Outcome.SingleAssetExit[] memory sourceOutcome,
+            Outcome.SingleAssetExit[] memory targetOutcome,
+            uint256 initialAssetHoldings
+        )
+    {
+        _requireIncreasingIndices(indices); // This assumption is relied on by compute_transfer_effects_and_interactions
+
+        // source checks
+        _requireChannelFinalized(sourceChannelId);
+        _requireMatchingFingerprint(
+            sourceStateHash,
+            keccak256(sourceOutcomeBytes),
+            sourceChannelId
+        );
+
+        sourceOutcome = Outcome.decodeExit(sourceOutcomeBytes);
+        targetOutcome = Outcome.decodeExit(targetOutcomeBytes);
+        address asset = sourceOutcome[assetIndex].asset;
+        require(targetOutcome[assetIndex].asset == asset, 'asset mismatch');
+        initialAssetHoldings = holdings[asset][sourceChannelId];
```

Comment:
I believe this check implies that the funding channel outcome should have the same assets (in the same order) as the virtually-funded (aka target) channel outcome. I believe that a single channel can fund many virtually-funded channels. In that case, every virtually-funded channel must contain every asset used by any other virtually-funded channel. 

I am not really recommending a change here, but just want to make sure my understanding is correct.

Reply:
I think your understanding is correct. I think it is worth avoiding that requirement if we can, so I'll take a stab at having both a `sourceAssetIndex` and `targetAssetIndex`.

Reply:
42dd3b6